core/src/main/scala/org/apache/spark/ui/scope/RDDOperationGraph.scala

```diff
@@ -165,12 +165,12 @@ private[ui] object RDDOperationGraph extends Logging {
    * For the complete DOT specification, see http://www.graphviz.org/Documentation/dotguide.pdf.
    */
   def makeDotFile(graph: RDDOperationGraph): String = {
-    val dotFile = new StringBuilder
-    dotFile.append("digraph G {\n")
-    dotFile.append(makeDotSubgraph(graph.rootCluster, indent = "  "))
-    graph.edges.foreach { edge => dotFile.append(s"""  ${edge.fromId}->${edge.toId};\n""") }
-    dotFile.append("}")
-    val result = dotFile.toString()
+    val subgraph = makeDotSubgraph(graph.rootCluster, indent = "  ")
+
+    val edges = new StringBuilder
```

Comment:
We don't want to use a single `StringBuilder` to concatenate all strings here. When  the `StringBuilder` appended the string of `makeDotSubgraph`, it may already contain a long string. If `graph.edges` has too many appending operations and it exceeds the `StringBuilder`'s capacity, `StringBuilder` will expands its capacity and ask twice memory. It can easily cause out-of-memory problem.